domain/secret/service/service_test.go

```diff
@@ -1500,6 +1500,57 @@ func (s *serviceSuite) TestWatchObsolete(c *gc.C) {
 	wC.AssertNoChange()
 }
 
+func (s *serviceSuite) TestWatchObsoleteUserSecretsToPrune(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	s.state = NewMockState(ctrl)
+	mockWatcherFactory := NewMockWatcherFactory(ctrl)
+
+	ch1 := make(chan struct{}, 1)
+	ch2 := make(chan struct{}, 1)
+
+	go func() {
+		ch1 <- struct{}{}
+		ch2 <- struct{}{}
+	}()
+
+	mockObsoleteWatcher := NewMockNotifyWatcher(ctrl)
+	mockObsoleteWatcher.EXPECT().Changes().Return(ch1).AnyTimes()
+	mockObsoleteWatcher.EXPECT().Wait().Return(nil).AnyTimes()
+	mockObsoleteWatcher.EXPECT().Kill().AnyTimes()
+
+	mockAutoPruneWatcher := NewMockNotifyWatcher(ctrl)
+	mockAutoPruneWatcher.EXPECT().Changes().Return(ch2).AnyTimes()
+	mockAutoPruneWatcher.EXPECT().Wait().Return(nil).AnyTimes()
+	mockAutoPruneWatcher.EXPECT().Kill().AnyTimes()
+
+	mockWatcherFactory.EXPECT().NewNamespaceNotifyMapperWatcher("secret_revision_obsolete", changestream.Create, gomock.Any()).Return(mockObsoleteWatcher, nil)
+	mockWatcherFactory.EXPECT().NewNamespaceNotifyMapperWatcher("secret_metadata", changestream.Update, gomock.Any()).Return(mockAutoPruneWatcher, nil)
+
+	svc := NewWatchableService(s.state, loggertesting.WrapCheckLog(c), mockWatcherFactory, nil)
+	w, err := svc.WatchObsoleteUserSecretsToPrune(context.Background())
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(w, gc.NotNil)
+	defer workertest.CleanKill(c, w)
+	wc := watchertest.NewNotifyWatcherC(c, w)
+	// initial change.
+	wc.AssertOneChange()
+
+	select {
+	case ch1 <- struct{}{}:
+	case <-time.After(coretesting.ShortWait):
+		c.Fatalf("timed out waiting for sending the secret revision changes")
+	}
+	select {
+	case ch2 <- struct{}{}:
+	case <-time.After(coretesting.ShortWait):
+		c.Fatalf("timed out waiting for sending the secret URI changes")
+	}
+
+	wc.AssertAtLeastOneChange()
```

Comment:
The test sends only 2 events so can we be more precise in what we are asserting.
To test the operation of the combined watcher, we should really send an event to each sub watcher and test that we get an event each time.